gateway/ln-gateway/src/gateway_module_ng/mod.rs

```diff
@@ -0,0 +1,287 @@
+            amount: input.amount,
+            fee: self.cfg.fee_consensus.contract_input,
+        })
+    }
+
+    fn output_amount(
+        &self,
+        output: &<Self::Common as ModuleCommon>::Output,
+    ) -> Option<TransactionItemAmount> {
+        Some(TransactionItemAmount {
+            amount: output.amount(),
+            fee: self.cfg.fee_consensus.contract_output,
+        })
+    }
+}
+
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Decodable, Encodable)]
+pub enum GatewayClientStateMachinesNG {
+    Send(SendStateMachine),
+    Receive(ReceiveStateMachine),
+}
+
+impl IntoDynInstance for GatewayClientStateMachinesNG {
+    type DynType = DynState;
+
+    fn into_dyn(self, instance_id: ModuleInstanceId) -> Self::DynType {
+        DynState::from_typed(instance_id, self)
+    }
+}
+
+impl State for GatewayClientStateMachinesNG {
+    type ModuleContext = GatewayClientContextNG;
+
+    fn transitions(
+        &self,
+        context: &Self::ModuleContext,
+        global_context: &DynGlobalClientContext,
+    ) -> Vec<StateTransition<Self>> {
+        match self {
+            GatewayClientStateMachinesNG::Receive(state) => {
+                sm_enum_variant_translation!(
+                    state.transitions(context, global_context),
+                    GatewayClientStateMachinesNG::Receive
+                )
+            }
+            GatewayClientStateMachinesNG::Send(state) => {
+                sm_enum_variant_translation!(
+                    state.transitions(context, global_context),
+                    GatewayClientStateMachinesNG::Send
+                )
+            }
+        }
+    }
+
+    fn operation_id(&self) -> OperationId {
+        match self {
+            GatewayClientStateMachinesNG::Receive(state) => state.operation_id(),
+            GatewayClientStateMachinesNG::Send(state) => state.operation_id(),
+        }
+    }
+}
+
+impl GatewayClientModuleNG {
+    pub async fn start_send_state_machine(
+        &self,
+        operation_id: OperationId,
+        max_delay: u64,
+        max_fee_msat: u64,
+        invoice: Bolt11Invoice,
+        contract: OutgoingContract,
+    ) -> anyhow::Result<()> {
+        let send_sm = GatewayClientStateMachinesNG::Send(SendStateMachine {
+            operation_id,
+            contract: contract.clone(),
+            state: SendSMState::Sending(Sending {
+                max_delay,
+                max_fee_msat,
+                invoice,
+                claim_keypair: self.keypair,
+            }),
+        });
+
+        self.client_ctx
+            .manual_operation_start(
+                operation_id,
+                LightningCommonInit::KIND.as_str(),
+                GatewayOperationMetaNG,
+                vec![self.client_ctx.make_dyn_state(send_sm)],
+            )
+            .await
+    }
+
+    pub async fn subscribe_send(
+        &self,
+        operation_id: OperationId,
+        contract: OutgoingContract,
```

Comment:
Since `OutgoingContract` is a parameter here, does this mean that in order for the gateway to subscribe to a contract, it needs to keep track of the `OutgoingContract` in addition to the `OperationId`? Can we put the `OutgoingContract` in the operation log so that we only need to keep track of the `OperationId`?